docker/recommenders/recommender_matrix_factorization.py

```diff
@@ -0,0 +1,110 @@
+import implicit
+import logging
+import numpy as np
+from scipy.sparse import coo_matrix
+from scipy.spatial.distance import cosine
+
+from recommenders.recommender import Recommender, _is_intersection
+from reader import event_to_tips
+
+logging.basicConfig(filename="recommendations.log", level=logging.INFO)
+
+
+class BayesianPersonalizedRanking(Recommender):
+    def _generate_indices(self):
+        self.device_to_index = {}
+        for i in range(len(self.train_devices)):
+            self.device_to_index[self.train_devices[i]] = i
+
+        self.event_to_index = {}
+        for i in range(len(self.event_types)):
+            self.event_to_index[self.event_types[i]] = i
+
+    def _generate_user_to_event_index(self):
+        self.user_to_event_index = {}
+        for event in self.train_events.keys():
+            device_id, group_id, event_id = event
+            if self.device_to_index[device_id] not in self.user_to_event_index.keys():
+                self.user_to_event_index[self.device_to_index[device_id]] = []
+            self.user_to_event_index[self.device_to_index[device_id]].append(self.event_to_index[(group_id, event_id)])
+
+    def _generate_matrix(self):
+        data = []
+        row_id = []
+        col_id = []
+        for event in self.train_events.keys():
+            (device_id, group_id, event_id) = event
+            data.append(1)
+            row_id.append(self.event_to_index[(group_id, event_id)])
+            col_id.append(self.device_to_index[device_id])
+
+        logging.info("BayesianPersonalizedRanking: matrix data computed")
+
+        matrix = coo_matrix((np.array(data), (np.array(row_id), np.array(col_id))),
+                            dtype=float, shape=(len(self.event_types), len(self.train_devices)))
+        return matrix
+
+    def __init__(self, train_devices, event_types, train_events):
+        super(BayesianPersonalizedRanking, self).__init__(train_devices, event_types, train_events)
+        
+        logging.info("BayesianPersonalizedRanking: init started")
+
+        self._generate_indices()
+
+        logging.info("BayesianPersonalizedRanking: device_to_index and event_to_index computed")
+
+        self._generate_user_to_event_index()
+         
+        self.matrix = self._generate_matrix()
```

Comment:
As I understood, we train a new model on every server start up. Is it possible to train the models only once, before deploying the server? It seems, we can train a model, then pickle it, and put into the docker image the pickled model. Thus, we don't need event log inside the docker image anymore. Also, that should significantly speed up server start up and alleviate cpu/memory requirements.

Reply:
Done.